src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoNotNullable.cs

```diff
@@ -127,7 +166,17 @@ internal override void WriteEnumerable(JsonSerializerOptions options, ref WriteS
             if (ValueConverter != null)
             {
                 Debug.Assert(current.Enumerator != null);
-                TRuntimeProperty value = (TRuntimeProperty)current.Enumerator.Current;
+
+                TRuntimeProperty value;
+                if (current.Enumerator is IEnumerator<TRuntimeProperty>)
+                {
+                    value = ((IEnumerator<TRuntimeProperty>)current.Enumerator).Current;
+                }
+                else
+                {
+                    value = (TRuntimeProperty)current.Enumerator.Current;
```

Comment:
Is this cast guaranteed to succeed?

Reply:
Yes. `TRuntimeProperty` is always going to be the same as `TDeclaredProperty` except in the case where we support the special `object` mode. In that case, we create a new property info class based on the each element.GetType().